Typed-endpoint-docs (#31166)
zulip/zulip: docs/tutorials/writing-views.md

  will check that the "subscriptions" HTTP parameter is a list of
  strings, marshalled as JSON, and pass it into the function with the
  Python keyword argument `streams_raw`.

- `message_id: Json[NonNegativeInt]` will check that the `message_id`
  HTTP parameter is a string containing a JSON encoded non-negative
  integer.

[Annotated](https://docs.python.org/3/library/typing.html#typing.Annotated)
can be used in combination with
[Pydantic's validators](https://docs.pydantic.dev/latest/api/functional_validators/)
to provide additional validation for the arguments.

- `name: Annotated[str, StringConstraints(max_length=60)]` will check that the
  `name` HTTP parameter is a string containing up to 60 characters.

- Since there is no need to JSON-encode strings
  (lists, integers, bools and complex objects require JSON encoding), usually simply
  `my_string: str` is correct. One can pass, for example,
  `Annotated[str, check_string_in_validator(...)]` where one wants to run a
  validator on the value of a string.

Default values can be specified for optional arguments similar to how we would specify
default values in regular python function.

> - `mandatory: Json[bool] = False` will assign False to the `mandatory` argument if

Reviewer: `mandatory` as the parameter name feels like a juxtaposition here. Maybe something more generic?

  no value is specified when making a request to the endpoint.

- We can use `None` as the default value for optional arguments when we don't
  want to specify any specific default value, for example,

>   `narrow: Json[list[NarrowParameter]] | None = None`.

Reviewer: Might as well mention that `Json[...] | None` does not allow `null` as a client provided value, which is the case for most of our optional request variables.

You can find some context on this [here](https://github.com/zulip/zulip/blob/2011e0df760cea52c31914e7b77d9b4e38e9ee74/zerver/tests/test_typed_endpoint.py#L539-L552) (probably no need to link this in the documentation).

```py
from zerver.decorator import require_realm_admin
from zerver.lib.typed_endpoint import typed_endpoint

@require_realm_admin
@typed_endpoint
def create_user_backend(
        request: HttpRequest,
        user_profile: UserProfile,
        *,
        email: str,
        password: str,
        full_name: str,
    ):
    # ... code here
```

> The `typed_endpoint` decorator parses the declared [keyword-only arguments](https://docs.python.org/3/glossary.html#term-parameter)

Reviewer 2: For docs, we generally want to keep each line under 76 chars in length. So, I think we can add a line break here

```py
from zerver.decorator import require_realm_admin
from zerver.lib.typed_endpoint import typed_endpoint

@require_realm_admin
@typed_endpoint
def create_user_backend(
        request: HttpRequest,
        user_profile: UserProfile,
        *,
        email: str,
        password: str,
        full_name: str,
    ):
    # ... code here
```

> The `typed_endpoint` decorator parses the declared keyword arguments

Reviewer: keyword -> keyword-only, and a link to Python's glossary could help: https://docs.python.org/3/glossary.html#term-parameter.

of the decorated function, and for each argument that has been declared,
it extracts the HTTP parameter with that name from the request,

> parses it as JSON(if the argument is declared as JSON), and passes it to

Reviewer: Missing a space between "JSON" and "(if the argument.."

However, looks like even in the original documentation, this line was a bit misguided as the example did not use `json_validator`.

I think simply saying something like `parses it according to the type annotation` suffices, and then we can bring up advanced details like `Json[...]` and friends later.

the function. It will return an nicely JSON formatted HTTP 400 error
in the event that an argument is missing, doesn't parse as JSON, or
otherwise is invalid.

`require_realm_admin` is another decorator which checks the
authorization of the given `user_profile` to make sure it belongs to a
realm administrator (and thus has permission to create a user); we
show it here primarily to show how `typed_endpoint` should be
the inner decorator.

The implementation of `typed_endpoint` is documented in detail
in
[zerver/lib/typed_endpoint.py](https://github.com/zulip/zulip/blob/main/zerver/lib/typed_endpoint.py)

Pydantic also helps us with request variable validation. For example:

- `msg_ids: Json[list[int]]` will check that the `msg_ids`
  HTTP parameter is a list of integers, marshalled as JSON,
  and pass it into the function as the `msg_ids` Python
  keyword argument.

> - `streams_raw: Annotated[Json[list[str]],ApiParamConfig("subscriptions")]`

Reviewer: Missing a space here
```suggestion
- `streams_raw: Annotated[Json[list[str]], ApiParamConfig("subscriptions")]`
```

  will check that the "subscriptions" HTTP parameter is a list of
  strings, marshalled as JSON, and pass it into the function with the
  Python keyword argument `streams_raw`.

- `message_id: Json[NonNegativeInt]` will check that the `message_id`
  HTTP parameter is a string containing a JSON encoded non-negative
  integer.

[Annotated](https://docs.python.org/3/library/typing.html#typing.Annotated)
can be used in combination with
[Pydantic's validators](https://docs.pydantic.dev/latest/api/functional_validators/)
to provide additional validation for the arguments.

- `name: Annotated[str, StringConstraints(max_length=60)]` will check that the

>   `name` HTTP parameter is a string containing of length less than 60 characters.

Reviewer: Because it is `max_length=60`, it would be more accurate to say:

```suggestion
  `name` HTTP parameter is a string containing up to 60 characters.
```

- Since there is no need to JSON-encode strings, usually simply
  `my_string: str` is correct. One can pass, for example,
  `Annotated[str, check_string_in_validator(...)]` where one wants to run a
  validator on the value of a string.

See
[zerver/lib/typed_endpoint_validators.py](https://github.com/zulip/zulip/blob/main/zerver/lib/typed_endpoint_validators.py)

> for more validators and their documentation.

Reviewer: Let's also demonstrate the use of default values here.
